frontend/src/workers/serviceWorker.ts

```diff
@@ -0,0 +1,208 @@
+/**
+ * At first, the SW is installed.
+ * Note that the SW is not yet running, as it needs to be activated before.
+ */
+self.addEventListener('install', event => {
+  // eslint-disable-next-line no-console
+  console.log('Attempting to install service worker')
+
+  event.waitUntil(
+    caches
+      .open(CACHE_NAME)
+      .then(cache =>
+        /**
+         * Add all static files to the cache
+         */
+        cache.addAll(STATIC_ASSETS)
+      )
+      // eslint-disable-next-line no-console
+      .catch(err => console.log(err))
+  )
+})
+
+/**
+ * After being installed and on page reload, the SW is activated
+ */
+self.addEventListener('activate', event => {
+  // eslint-disable-next-line no-console
+  console.log('Activating new service worker...')
+
+  event.waitUntil(
+    caches.keys().then(cacheNames => {
+      /**
+       * The claim() method causes those pages to be controlled immediately - Use fetch event handling right away.
+       * @see https://developer.mozilla.org/en-US/docs/Web/API/Clients/claim
+       */
+      // @ts-ignore
+      self.clients.claim()
+
+      /**
+       * If a new service worker is activated, it will delete the old cache entries
+       */
+      return Promise.all(
+        cacheNames.map(storedCacheName => {
+          if (storedCacheName !== CACHE_NAME) {
+            return caches.delete(storedCacheName)
+          }
+
+          return Promise.reject()
+        })
+      )
+    })
+  )
+})
+
+/**
+ * Custom message used by the application.
+ *
+ * Types are:
+ * - CACHED_REQUEST_SIZE:
+ *      Get the size of all cached requests
+ * - UPDATE_CACHE:
+ *      This message us used to write and update the WHITELISTED_AND_READ_ONLY_PATHS.
+ *      These paths caches can't be updated by the `fetch()` method.
+ */
+self.addEventListener('message', async event => {
+  if (event.data === CACHED_REQUEST_SIZE) {
+    const cache = await caches.open(CACHE_NAME)
+    const requests = await cache.keys()
+
+    // @ts-ignore
+    event.source?.postMessage({
+      data: requests.length,
+      type: CACHED_REQUEST_SIZE
+    })
+  }
+
+  if (event.data === UPDATE_CACHE) {
+    const cache = await caches.open(CACHE_NAME)
+
+    try {
+      await cache.delete(new Request(INFRACTIONS))
+      await cache.add(INFRACTIONS)
+    } catch (e) {
+      // @ts-ignore
+      event.source?.postMessage({
+        data: e,
+        type: UPDATE_CACHE
+      })
+
+      return
+    }
+
+    try {
+      await cache.delete(new Request(REGULATIONS))
+      await cache.add(REGULATIONS)
+    } catch (e) {
+      // @ts-ignore
+      event.source?.postMessage({
+        data: e,
+        type: UPDATE_CACHE
+      })
+
+      return
+    }
+
+    // @ts-ignore
+    event.source?.postMessage({
+      type: UPDATE_CACHE
+    })
+  }
+})
+
+/**
+ * Entrypoint - when fetching an asset or an API
+ */
+self.addEventListener('fetch', event => {
+  event.respondWith(
+    (async () => {
+      const url = event.request.url.toString()
```

Comment:
TODO for me : params, hashes ?

Reply:
Il y a tout ce qui est dans l'URL : https://developer.mozilla.org/en-US/docs/Web/API/Request/url
On n'a pas de hashes dans les URLs cachés